rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,114 @@
+---
+feature: Defined conventions around testing of official Nixpkgs packages.
+start-date: 2021-09-03
+author: Jonathan Ringer
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+  - [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88)
+---
+
+# Summary
+[summary]: #summary
+
+When updating or modifying packages, several conventions for testing regressions
+have been adopted. However, these practices are not standard, and generally it's not well
+defined how each testing method should be implemented. It would be beneficial to have
+an unambiguous way to say that a given package, and all downstream dependencies, have
+had as many automated test ran possible. This will give a high degree of certainty that
+a given change is less likely to manifest regressions once introduced on a release
+channel.
+
+Another desire of this rfc is also to have a way for various review tools
+(e.g. ofborg, hydra, nixpkgs-review) to have a standard way to determine if a
+package has additional tests which can help verify it's correctness.
+
+# Motivation
+[motivation]: #motivation
+
+Breakages are a constant painpoint for nixpkgs. It is a very poor user experience to
+have a configuration broken because one or more packages fail to build. Often when
+these breakages occur, it is because the change had a large impact on the entirity
+of nixpkgs; and unless there's a dedicated hydra jobset for the pull request, it's
+infeasible to expect pull request authors to verify every package affected
+by a change they are proposing. However, it is feasible to specify packages that
+are very likely to be affected by changes in another package, and use this information
+to help mitigate regressions from appearing in release channels.
+
+# Detailed design
+[design]: #detailed-design
+
+Standardize `passthru.tests.<name>` and `passthru.nixosTests.<name>` as a mechanism of 
+more expensive but automatic testing for nixpkgs. As well as encourage the usage of
+`checkPhase` or `installCheckPhase` when packaging within nixpkgs.
```

Comment:
I'm confused where to draw the line between what should be done in `installCheckPhase` and non-vm `passthru.tests`. When reviewing this has been a source of confusion, see: https://github.com/NixOS/nixpkgs/pull/153496 as an example. (Reading this I'm no longer convinced the decision to move it from installCheckPhase was the right one :>)


cc @thiagokokada so that you can add your thoughts

Reply:
Also, IMO, VM-less `passthru.tests` makes more sense for tests like the one described in this document: maybe we want to see how this particular package being build interacts with other packages. But if the idea is test only the current package, `installCheckPhase` generally makes more sense to me (of course, there could be exceptions).

Reply:
I think ideally we want expensive build steps spread among different derivations, but that's hard. Expensive tests are sometimes feasible to split, so splitting them is good when it doesn't create a mess.

(Cheap checks with no new deps do not follow this logic, of course)

Also, separated tests make it more practical to support packages with different level of «works» on different platforms, because we do not need to touch the main build causing  expensive rebuild on the «happy» platforms to add partial support (and its tests) on the more exotic ones. There is no single «package is broken», in the black-and-white view we have never had a non-broken LibreOffice packages, for example.

Reply:
Separated tests also make it easier to debug failures, since you can easily access the package (which may build successfully but then fail the test). I think we should follow something like a "principle of least privilege", where it is best practice to always use the test that makes the fewest assumptions about its environment.

`checkPhase` assumes it is run within the build environment. That's a lot of assumptions. This increases build time for quick iteration, makes it harder to debug failures and also makes the test results less meaningful since the environment differs from the end-user's environments.

`installCheckPhase` has somewhat fewer assumptions than `checkPhase`, but still many of the same disadvantages. The results are more meaningful though, since the environment is more similar to the end-user's environment.

`passthru.tests` makes the fewest assumptions. If the tests pass in this "clean environment", they will likely pass in the user's environment as well (unless tests are actually broken by some environment variable or something similar). They can be run as-needed (but should be run (semi-)automatically in some reliable fashion). They make debugging easier.

`passthru.nixosTests` assumes access to a nixos VM, but does not assume a special build environment.

As a result, I'd suggest to pick `passthru.tests` > `passthru.nixosTests` > `installCheckPhase` > `checkPhase`. Practicality is king of course: If upstream assumptions make it trivial to run the test suite in `checkPhase` but hard to run them in `passthru.tests`, then its better to have some tests in `checkPhase` than to have no tests at all.

Reply:
I don't have much ~ecosystem perspective on this, but I do have a few scattered thoughts from authoring and maintaining a small number of packages (broadly, my thinking has been aligning with timokau over time):

1. I like `passthru.tests` at the front of the list. I can think of at least a few times I've been in the middle of some meaningful work and ended up derailed by an immediate or remote test or test-dependency failure:

    - I've started to see a check/test dependency on a linter or formatter as an antipattern in my own work. At least twice I've had a linter/formatter package turn up broken, or a bump in nixpkgs to trigger a rebuild that fails due to a lint/format check.
    
    - Recently I noticed that resholve's CLI arg parser dependency wasn't building on aarch64-darwin due to some failing tests. This was a pain since it was a platform I don't have readily available. I ended up overriding the dependency to disable checks for expediency, since resholve's own test suite (which I factored out into `passthru.tests` a while back) already exercises the CLI and should fail on borg/hydra if it did ultimately depend on whatever caused the test breaks.
    
2. I've thought for a while that it would be nice to have a low-friction mechanism for collecting example uses (say, a nudge to submit them + a CLI/form/template for doing so? Examples themselves might just be a valid invocation of some CLI, a sample file that exercises some parser, etc.) from package users. It occurs to me now that if there was a ~standard `passthru.tests.x` attr for these, it could at least drive the nudge (i.e. example, if you open a nix-shell with something that has fewer than N?)
3. I like prioritizing `installCheckPhase` over `checkPhase`. I have some minor insecurity about `checkPhase` providing false-confidence about something that `installPhase` or `fixupPhase` then breaks (especially since resholve presently runs in fixup).